Integrate API for displaying featured images

maskaravivek opened the pull request:
## Description

Part of task described in #324.

## Tests performed
Please test on `prodDebug` build. Apparently, the beta site doesn't have any featured images. 

## Screenshots showing what changed
 
![featured_images](https://user-images.githubusercontent.com/3069373/39017910-2bd5bdf8-4442-11e8-8fac-e881ebbc69f7.png)

misaochan commented on app/build.gradle:
```diff
@@ -48,6 +48,8 @@ dependencies {
     implementation 'com.jakewharton.rxbinding2:rxbinding-appcompat-v7:2.0.0'
     implementation 'com.jakewharton.rxbinding2:rxbinding-design:2.0.0'
 
+    implementation 'org.jsoup:jsoup:1.11.3'
```
Hi @maskaravivek , what is the reason for needing to add this library?

misaochan commented:
Thanks for this PR, @maskaravivek ! It works well for me in general, great job. A few issues:

- It only loads 10 images, and the same images all the time. This will get boring very quickly for the user, and they will likely not bother to give this feature more than one cursory glance. I think that for our minimum viable product, we can load as many images as the user's "Set Recent Upload Limit" setting states? I know that that is meant for contributions, but IMO realistically if they can load 100 of their own images, they can load 100 of other people's images without any issue. Ideally we should randomize the pictures shown or dynamically load more after scrolling down, but that can be an enhancement for the future.
- In the media details view (accessed when image is tapped), the author field says "Featured image author user name goes here". :)
- The toast "Error! URL not found" is displayed when the media details view loads, even though the image and most of the fields are correct
- After viewing a few images, I crashed with the error:

```
04-30 21:31:33.051 23958-23958/fr.free.nrw.commons.debug E/AndroidRuntime: FATAL EXCEPTION: main
    Process: fr.free.nrw.commons.debug, PID: 23958
    java.util.ConcurrentModificationException
        at java.util.HashMap$HashIterator.nextEntry(HashMap.java:851)
        at java.util.HashMap$KeyIterator.next(HashMap.java:885)
        at fr.free.nrw.commons.Media.setDescriptions(Media.java:359)
        at fr.free.nrw.commons.MediaDataExtractor.fill(MediaDataExtractor.java:303)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.onPostExecute(MediaDetailFragment.java:256)
        at fr.free.nrw.commons.media.MediaDetailFragment$3.onPostExecute(MediaDetailFragment.java:229)
        at android.os.AsyncTask.finish(AsyncTask.java:667)
        at android.os.AsyncTask.-wrap1(AsyncTask.java)
        at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:684)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:154)
        at android.app.ActivityThread.main(ActivityThread.java:6119)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:886)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:776)
```

maskaravivek commented:
@misaochan Thanks for reviewing and testing the PR. 

I have fixed most of the issues reported by you. 
- "Error! URL not found" still isn't fixed since the `MediaDetailFragment` fetches the media details again from the server and is not able to fetch the license information. Thus the toast is displayed. Later on, we can pick refactoring of this fragment so that the data is fetched again only if there are missing data. 
- Have also implemented pagination here so that the user can scroll through as many images as he wishes. We can introduce randomization later on. 
- Have completely refactored the code so that it can be reused for displaying category images. 

@ujjwalagrawal17 @neslihanturan Just adding you guys to the conversation so that you don't have to rewrite the whole thing when/if browsing images of a particular category are picked up. 

Here's how to use it: 
```
CategoryImagesActivity.startYourself(this, <page_title>, <category_name>);
```

misaochan commented:
Hi @maskaravivek ,

- The pagination works very well! :) Nice work.
- The author field seems to disappear completely now when I load an image's details. Sorry to nitpick, but I think we really need to have an author field if we are displaying other people's images, especially as some of them have licenses that require attribution.
- Just noticed this:  Pressing the "back" button on the device closes the Commons app entirely at the moment, whereas it should be bringing users back to Contributions activity. (Actually, even going to "Home" from the nav drawer on Featured activity closes the app, which isn't intended)

misaochan commented on app/src/main/java/fr/free/nrw/commons/category/CategoryImageUtils.java:
```diff
@@ -0,0 +1,146 @@
+
+    private static Date getDateCreated(Node document) {
+        String dateTime = getMetaDataValue(document, "DateTime");
+        Timber.d("Date time is %s", dateTime);
```
Could you please make this verbose at least? It spams my entire logcat screen each time any page loads. :)

maskaravivek replied:
Will remove it altogether. Was just for debugging purposes :D

maskaravivek commented:
@misaochan The app closing on pressing back isn't happening on API 19 but am able to reproduce it on API 25. :(

maskaravivek commented:
@misaochan I have fixed all the issues reported by you. Thanks again for helping with testing. :)

misaochan commented:
@maskaravivek Does the author field work for you currently?

maskaravivek commented:
@misaochan Have fixed the travis build. :)

misaochan commented:
The build was re-run 5 min ago, but it seems to still fail:

```
13 PMD rule violations were found. See the report at: file:///home/travis/build/commons-app/apps-android-commons/app/build/reports/pmd/pmd.html
:app:kaptGenerateStubsBetaDebugUnitTestKotlin
:app:kaptBetaDebugUnitTestKotlin
:app:compileBetaDebugUnitTestKotlin
:app:preBetaDebugUnitTestBuild UP-TO-DATE
:app:javaPreCompileBetaDebugUnitTest
:app:compileBetaDebugUnitTestJavaWithJavac NO-SOURCE
:app:mockableAndroidJar
:app:processBetaDebugJavaRes
:app:processBetaDebugUnitTestJavaRes NO-SOURCE
:app:testBetaDebugUnitTest
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > editToken FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > simpleLogin FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > simpleLoginWithWrongPassword FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > validateLoginForLoggedInUser FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > twoFactorLogin FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > validateLoginForLoggedOutUser FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > getUploadCount FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > fileExistsWithName_FileNotFound FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
fr.free.nrw.commons.mwapi.ApacheHttpClientMediaWikiApiTest > authCookiesAreHandled FAILED
    kotlin.UninitializedPropertyAccessException at ApacheHttpClientMediaWikiApiTest.kt:36
112 tests completed, 9 failed
:app:testBetaDebugUnitTest FAILED
FAILURE: Build failed with an exception.
* What went wrong:
Execution failed for task ':app:testBetaDebugUnitTest'.
> There were failing tests. See the report at: file:///home/travis/build/commons-app/apps-android-commons/app/build/reports/tests/testBetaDebugUnitTest/index.html
* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.
* Get more help at https://help.gradle.org
BUILD FAILED in 6m 28s
76 actionable tasks: 76 executed
```

neslihanturan commented:
@maskaravivek , it gives me this screen in a loop:
![loop](https://user-images.githubusercontent.com/3127881/39572463-e957b86e-4ed7-11e8-97ac-2a3d8eb4fde4.png)

maskaravivek commented:
@neslihanturan Can you post the error logs from logcat. Are you testing on beta or prod?

maskaravivek commented:
@misaochan Finally the build succeeds. Sorry for having you test it multiple times.

misaochan commented:
Thanks @maskaravivek . All good on my end - we can merge after @neslihanturan 's issue is sorted. :)

neslihanturan commented:
Sorry for not adding logs earlier, here:
` Error occurred while loading featured images
                                                                                  java.lang.NullPointerException: Attempt to invoke virtual method 'void fr.free.nrw.commons.category.GridViewAdapter.addItems(java.util.List)' on a null object reference
                                                                                      at fr.free.nrw.commons.category.CategoryImagesListFragment.lambda$fetchMoreImages$4$CategoryImagesListFragment(CategoryImagesListFragment.java:127)
                                                                                      at fr.free.nrw.commons.category.CategoryImagesListFragment$$Lambda$4.accept(Unknown Source)
                                                                                      at io.reactivex.internal.observers.LambdaObserver.onNext(LambdaObserver.java:60)
                                                                                      at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.drainNormal(ObservableObserveOn.java:200)
                                                                                      at io.reactivex.internal.operators.observable.ObservableObserveOn$ObserveOnObserver.run(ObservableObserveOn.java:252)
                                                                                      at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run(HandlerScheduler.java:109)
                                                                                      at android.os.Handler.handleCallback(Handler.java:739)
                                                                                      at android.os.Handler.dispatchMessage(Handler.java:95)
                                                                                      at android.os.Looper.loop(Looper.java:148)
                                                                                      at android.app.ActivityThread.main(ActivityThread.java:5417)
                                                                                      at java.lang.reflect.Method.invoke(Native Method)
                                                                                      at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:726)
                                                                                      at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:616)`

maskaravivek commented:
@neslihanturan I have added checks to gracefully handle the error cases. 

![no_images](https://user-images.githubusercontent.com/3069373/39661499-fcc15696-506f-11e8-874e-3794b7aab5d1.png)
![no_internet](https://user-images.githubusercontent.com/3069373/39661500-fd7ea958-506f-11e8-8fcc-85d74c651a57.png)

misaochan commented:
Thanks for adding the Javadocs! :) @neslihanturan I would like to merge this once you can confirm that the error handling works for you.

neslihanturan commented:
Works as expected, thanks you to include a great feature to our project @maskaravivek :)
